File: src/Wc3FrameExtLua/Fdf/Simple/String.lua
--=========
-- Include
--=========

local Class = LibManager.getDepency('LuaClass') or error('')
---@type Wc3Utils
local Wc3Utils = LibManager.getDepency('Wc3Utils') or error('')
local isTypeErr = Wc3Utils.isTypeErr or error('')
local Log = Wc3Utils.Log or error('')

---@type FdfDataClass
local FdfData = require('FdfData') or error('')

--=======
-- Class
--=======

local FdfSimpleString = Class.new('FdfSimpleString', FdfData)
---@class FdfSimpleString : FdfData
local public = FdfSimpleString.public
---@class FdfSimpleStringClass : FdfDataClass
local static = FdfSimpleString.static
---@type FdfSimpleStringClass
local override = FdfSimpleString.override
local private = {}

--=========
-- Static
--=========

---@param name string
---@param child FdfSimpleString | nil
---@return FdfSimpleString
function override.new(name, child)
    isTypeErr(name, 'string', 'name')
    if child then isTypeErr(child, FdfSimpleString, 'child') end

    local instance = child or Class.allocate(FdfSimpleString)
    instance = FdfData.new(name, 'String', true, nil, instance)

    return instance
end

--========
-- Public
--========

---@param path string
---@param size number
function public:setFont(path, size)
    isTypeErr(self, FdfSimpleString, 'self')
    isTypeErr(path, 'string', 'path')
    isTypeErr(size, 'number', 'size')

    self:setParameter('Font', '\"'..path..'\", '..tostring(size))
end

--- [0,1]
---@param r number
---@param g number
---@param b number
---@param a number
function public:setColor(r, g, b, a)
    isTypeErr(self, FdfSimpleString, 'self')
    isTypeErr(r, 'number', 'r')
    isTypeErr(g, 'number', 'g')
    isTypeErr(b, 'number', 'b')
    isTypeErr(a, 'number', 'a')

    r = r < 0 and 0 or r > 1 and 1 or r
    g = g < 0 and 0 or g > 1 and 1 or g
    b = b < 0 and 0 or b > 1 and 1 or b
    a = a < 0 and 0 or a > 1 and 1 or a

    self:setParameter('FontColor', tostring(r)..' '
                                 ..tostring(g)..' '
                                 ..tostring(b)..' '
                                 ..tostring(a))
end

---@param text string
function public:setText(text)
    isTypeErr(self, FdfSimpleString, 'self')
    isTypeErr(text, 'string', 'text')

    self:setParameter('Text', '\"'..text..'\"')
end

function public:addSubframe()
    isTypeErr(self, FdfSimpleString, 'self')

    Log:err(tostring(FdfSimpleString)..' can not have subframes')
end

---@return string
function public:serialize()
    isTypeErr(self, FdfSimpleString, 'self')

    local res = string.format("%s \"%s\" {\n", self:getBaseType(), self:getName())
    for param, value in pairs(self:getAllParameters()) do
        res = res..'    '..param..' '..value..',\n'
    end
    return res.."}"
end

--=========
-- Private
--=========

return static

File: src/FrameExt/Fdf/Simple/String.ts
import { Color, Log } from "../../../Utils";
import { Fdf } from "../../Fdf";

export class FdfSimpleString extends Fdf {
    constructor(name: string){
        super(name, 'String', true)
    }

    public get text(){return this._text}
    public set text(text: string){
        this._setParam('Text', '\"' + text + '\"')
        this._text = text
    }

    public get font(){return this._font}
    public set font(path: string){
        this._setParam('Font', '\"' + path + '\", ' + this._font_size.toString())
        this._font = path
    }

    public get fontSize(){return this._font_size}
    public set fontSize(size: number){
        this._setParam('Font', '\"' + this._font + '\", ' + size.toString())
        this._font_size = size
    }

    public get color(){return new Color(this._color)}
    public set color(c: Color){
        this._setParam('FontColor', string.format('%f %f %f %f', c.r, c.g, c.b, c.a))
        this._color = new Color(c)
    }

    public addSubframe(){
        return Log.err(FdfSimpleString.name + 
                       ': can not have subframes.', 2)
    }
    public getSubframe(){
        return Log.err(FdfSimpleString.name + 
                       ': can not have subframes.', 2)
    }
    public removeSubframe(){
        return Log.err(FdfSimpleString.name + 
                       ': can not have subframes.', 2)
    }

    public serialize(){
        let res = string.format("%s \"%s\" {\n", this.base_type, this.name)
        for (let [param, value] of this._parameters){
            res += '    ' + param + ' ' + value + ',\n'
        }
        res += '}'
        return res
    }

    private _text: string = ''
    private _font: string = ''
    private _font_size: number = 0.01
    private _color: Color = new Color(1, 1, 1, 1)
}